File: targets/widget/ApiClient.swift
import Foundation

enum FetchOutcome<T> {
  case success(T)
  case noData   // reachable, but the requested slice is null / empty
  case failure  // network or auth error
}

/// Minimal evcc API client (GET jq slices of /api/state + POST actions), with
/// optional basic auth.
enum ApiClient {
  private static func base(_ server: StoredServer) -> String {
    var b = server.url
    while b.hasSuffix("/") { b.removeLast() }
    return b
  }

  private static func authorize(_ req: inout URLRequest, _ server: StoredServer) {
    if server.authRequired == true, let u = server.username, !u.isEmpty, let p = server.password {
      let token = Data("\(u):\(p)".utf8).base64EncodedString()
      req.setValue("Basic \(token)", forHTTPHeaderField: "Authorization")
    }
    // Cloudflare Access service token (stateless, no cookie handling needed)
    if let id = server.serviceTokenId, !id.isEmpty,
       let secret = server.serviceTokenSecret, !secret.isEmpty {
      req.setValue(id, forHTTPHeaderField: "CF-Access-Client-Id")
      req.setValue(secret, forHTTPHeaderField: "CF-Access-Client-Secret")
    }
  }

  static func request(_ server: StoredServer, jq: String) -> URLRequest? {
    guard var comps = URLComponents(string: base(server) + "/api/state") else { return nil }
    comps.queryItems = [URLQueryItem(name: "jq", value: jq)]
    guard let url = comps.url else { return nil }
    var req = URLRequest(url: url, timeoutInterval: 15)
    req.cachePolicy = .reloadIgnoringLocalCacheData
    authorize(&req, server)
    return req
  }

  static func fetch<T: Decodable>(_ server: StoredServer, jq: String, as type: T.Type) async -> FetchOutcome<T> {
    guard let req = request(server, jq: jq) else { return .failure }
    do {
      let (data, response) = try await URLSession.shared.data(for: req)
      guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
        return .failure
      }
      if let s = String(data: data, encoding: .utf8)?
        .trimmingCharacters(in: .whitespacesAndNewlines),
        s.isEmpty || s == "null" || s == "[]" || s == "{}" {
        return .noData
      }
      let decoded = try EvccJSON.decoder.decode(T.self, from: data)
      return .success(decoded)
    } catch {
      return .failure
    }
  }

  /// POST to an API path (e.g. "/api/loadpoints/1/mode/pv"). Returns success.
  @discardableResult
  static func post(_ server: StoredServer, path: String) async -> Bool {
    let p = path.hasPrefix("/") ? path : "/" + path
    guard let url = URL(string: base(server) + p) else { return false }
    var req = URLRequest(url: url, timeoutInterval: 15)
    req.httpMethod = "POST"
    authorize(&req, server)
    do {
      let (_, response) = try await URLSession.shared.data(for: req)
      return (response as? HTTPURLResponse).map { (200..<300).contains($0.statusCode) } ?? false
    } catch {
      return false
    }
  }
}

File: targets/android-widget/kotlin/ApiClient.kt
package io.evcc.android.widget

import android.util.Base64
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLEncoder

sealed interface FetchOutcome {
    data class Success(val json: String) : FetchOutcome
    object NoData : FetchOutcome   // reachable, but the jq slice is null / empty
    object Failure : FetchOutcome  // network or auth error
}

/**
 * Minimal evcc API client: GET jq slices of /api/state and POST actions, with
 * optional basic auth. Android counterpart of ApiClient.swift. Runs on a
 * background thread (call from a coroutine / worker).
 */
object ApiClient {
    private const val TIMEOUT_MS = 15_000

    private fun base(server: StoredServer): String = server.url.trimEnd('/')

    private fun authorize(conn: HttpURLConnection, server: StoredServer) {
        if (server.authRequired && !server.username.isNullOrEmpty() && server.password != null) {
            val token = Base64.encodeToString(
                "${server.username}:${server.password}".toByteArray(),
                Base64.NO_WRAP,
            )
            conn.setRequestProperty("Authorization", "Basic $token")
        }
    }

    /** GET /api/state?jq=<jq>. Returns the raw response body on success. */
    fun fetch(server: StoredServer, jq: String): FetchOutcome {
        val url = "${base(server)}/api/state?jq=" + URLEncoder.encode(jq, "UTF-8")
        return runCatching {
            val conn = (URL(url).openConnection() as HttpURLConnection).apply {
                connectTimeout = TIMEOUT_MS
                readTimeout = TIMEOUT_MS
                requestMethod = "GET"
                authorize(this, server)
            }
            try {
                if (conn.responseCode !in 200..299) return FetchOutcome.Failure
                val body = conn.inputStream.bufferedReader().use { it.readText() }.trim()
                if (body.isEmpty() || body == "null" || body == "[]" || body == "{}") {
                    FetchOutcome.NoData
                } else {
                    FetchOutcome.Success(body)
                }
            } finally {
                conn.disconnect()
            }
        }.getOrDefault(FetchOutcome.Failure)
    }

    /** Loadpoint titles for the widget config picker, index-aligned to .loadpoints[]. */
    fun loadpointTitles(server: StoredServer): List<String> {
        val out = fetch(server, "[.loadpoints[].title]")
        if (out !is FetchOutcome.Success) return emptyList()
        return runCatching {
            val arr = org.json.JSONArray(out.json)
            (0 until arr.length()).map { i ->
                arr.optString(i).takeIf { it.isNotEmpty() && it != "null" } ?: "Loadpoint ${i + 1}"
            }
        }.getOrDefault(emptyList())
    }

    /** POST to an API path, e.g. "/api/loadpoints/1/mode/pv". Returns success. */
    fun post(server: StoredServer, path: String): Boolean {
        val p = if (path.startsWith("/")) path else "/$path"
        return runCatching {
            val conn = (URL(base(server) + p).openConnection() as HttpURLConnection).apply {
                connectTimeout = TIMEOUT_MS
                readTimeout = TIMEOUT_MS
                requestMethod = "POST"
                authorize(this, server)
            }
            try {
                conn.responseCode in 200..299
            } finally {
                conn.disconnect()
            }
        }.getOrDefault(false)
    }
}

/** Subset of /api/state .loadpoints[] used by the widget (see Loadpoint.swift). */
data class Loadpoint(
    val title: String?,
    val vehicleTitle: String?,
    val vehicleSoc: Double?,
    val effectiveLimitSoc: Double?,
    val chargePower: Double?,
    val mode: String?,
    val charging: Boolean,
    val connected: Boolean,
    val enabled: Boolean,
) {
    companion object {
        fun parse(json: String): Loadpoint? = runCatching {
            val o = JSONObject(json)
            fun d(k: String) = if (o.has(k) && !o.isNull(k)) o.optDouble(k) else null
            Loadpoint(
                title = o.optString("title").takeIf { it.isNotEmpty() },
                vehicleTitle = o.optString("vehicleTitle").takeIf { it.isNotEmpty() },
                vehicleSoc = d("vehicleSoc"),
                effectiveLimitSoc = d("effectiveLimitSoc"),
                chargePower = d("chargePower"),
                mode = o.optString("mode").takeIf { it.isNotEmpty() },
                charging = o.optBoolean("charging", false),
                connected = o.optBoolean("connected", false),
                enabled = o.optBoolean("enabled", false),
            )
        }.getOrNull()
    }
}
